src/main/java/io/strimzi/kafka/bridge/mqtt/kafka/BridgeKafkaProducerFactory.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright Strimzi authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+package io.strimzi.kafka.bridge.mqtt.kafka;
+
+import io.strimzi.kafka.bridge.mqtt.config.KafkaConfig;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Factory for creating Kafka producers as needed.
+ */
+public class BridgeKafkaProducerFactory<K, V> {
```

Comment:
not sure how the factory approach applies here. With the Factory pattern you simplify the way to build object of a different shapes but implementing the same interface. In this case it would be more having a `BridgeKafkaProducer` interface and then two concrete classes `BridgeKafkaProducerZero` and `BridgeKafkaProducerOne` which doesn't need an `ack` parameter in the constructor. Then the factory could also have just static methods to build each of them.
Other than this, a factory should not take responsability to hold the reference to what it builds and being a singleton to do so for sharing the producers across the application.
Let's try to think a better approach here.

Reply:
While the name suggests that it is a Factory, I would not say it is using exactly the same factory concept. I was just relying on some of its principle, and I came to this naive solution so far. 
To start, I was thinking of ways to get the specific producer (with ACK 1 or 0) without creating a producer every time a new MQTT message is received. 
That's why I came with this factory look-alike component, it just receives the config that will be shared across the producers, decide when to create the producers(only once), and the getProducer(QoS). 

Yes, I am considering the points you mentioned, and I will think about something better.

Reply:
Also take into account, as we saw for the loader class, that testing a singleton is not that simple, so if possible I would avoid it.

> To start, I was thinking of ways to get the specific producer (with ACK 1 or 0) without creating a producer every time a new MQTT message is received.

Absolutely. We don't need one producer per MQTT client/message.

Reply:
I had a hard time testing singleton :(
I need to investigate more and come with a good solution.